avocado_vt/test.py

```diff
@@ -405,17 +405,18 @@ def _runTest(self):
             test_modules[t_type] = imp.load_module(t_type, f, p, d)
             f.close()
 
-        # TODO: the environment file is deprecated code, and should be removed
-        # in future versions. Right now, it's being created on an Avocado temp
-        # dir that is only persisted during the runtime of one job, which is
-        # different from the original idea of the environment file (which was
-        # persist information accross virt-test/avocado-vt job runs)
-        env_filename = os.path.join(data_dir.get_tmp_dir(),
-                                    params.get("env", "env"))
+        # Open the environment file
+        custom_env_path = params.get("custom_env_path", "")
+        if custom_env_path:
+            env_path = custom_env_path
+        else:
+            env_path = data_dir.get_tmp_dir()
```

Comment:
IMHO, it's a new feature for VT and that means now `tmp_dir` is configurable, if so, we should implement it in `data_dir.get_tmp_dir()`, otherwise all the callers should implement the same logic and that would be horrible because in some context we don't have `params` been passed.

There could be two approaches:

1. introduce a new environment variable (e.g. `AVOCADO_VT_TMP_DIR`) and reference it in `data_dir.get_tmp_dir()`, then if someone wants to use a specified directory the avocado command line would be:  `AVOCADO_VT_TMP_DIR=/path/to/users/tmp avocado run ...`
2. introduce a new option to the VT configuration (https://github.com/avocado-framework/avocado-vt/blob/master/avocado_vt/conf.d/vt.conf), then users are only required to update the configuration and the avocado command line won't be changed. 

What do you think?

Reply:
Makes sense mostly because we could possibly reuse a persistent tmp path for other purposes later on or in all circumstances when we want to avoid the automated cleanup in between tests. I think a config item for avocado VT is more reasonable than an environment variable in order to keep all configuration in one place and a place that is more in the domain of a python program. I will ping you once I push code adding this change.

Reply:
However, why exactly do we have to add it to `data_dir.get_tmp_dir()`? The people would expect a true temporary directory instead of a persistent one. The directory obtained via `data_dir.get_tmp_dir()` is exactly the one that is wiped by the VT code after each test and this is exactly what we want to avoid by adding a parameter (or config or environmental variables) for a *persistent* tmp path. I guess we could still add the persistent tmp path as a config parameter or do you think then it is better off staying as Cartesian config parameter as we don't expect it to be used beyond the current environment processing scope?

Reply:
Hi @luckyh sorry for multiple edits on this comment - it has been quite many months and I constantly needed to refresh my mind on the changes in order to reply to you better. There are some troubles from your suggestion:

1) Changing the behavior of `data_dir.get_tmp_dir()` would affect way more people and code using it and the current overwriting with an optional temporary location is only persistent with respect to vm data across test runs, nothing else. This means that if for instance I set a config item "persistent_path" and change the `get_tmp_dir()` as:

```
    persistent_tmp_dir = settings.get_value('vt.setup', 'persistent_tmp_dir', default=None)
    if persistent_tmp_dir is None:
        return persistent_tmp_dir
```
This will replace the default temporary location with a custom one for all calling instances *while* this PR is only making sure the vm objects (running vms) are persistent, leaving the default temporary directory for everything else. We could call `get_tmp_dir(persistent=True)` only from the current locations but this brings me to the second point.

2) The changed code locations don't necessarily apply the same logic. While some (more than one) will check for persistent location first and if not will default to the original temporary directory, some have to additionally check if persistent location is set (in a boolean fashion) and this will require at least two lines of importing `settings` like above at each location which is about the same number of lines code we tried to reduce by refactoring into `get_tmp_dir`.

3) We also have other parameters used strictly for persistent vm objects like "vm_action" which are also Cartesian parameters like the "persistent_tmp_path". All of these parameters are Cartesian since this is the normal way to operate in `env_process.py` and `qemu_vm.py` and after some search I couldn't find a single instance of importing `avocado.settings` or `avocado_vt.options` into any module in the entire `virttest` folder.

Reply:
> 1. Changing the behavior of `data_dir.get_tmp_dir()` would affect way more people and code using it and the current overwriting with an optional temporary location is only persistent with respect to vm data across test runs, nothing else. This means that if for instance I set a config item "persistent_tmp_dir" and change the `get_tmp_dir()` as:
> 
> 
> ```
>     persistent_tmp_dir = settings.get_value('vt.setup', 'persistent_tmp_dir', default=None)
>     if persistent_tmp_dir is not None:
>         return persistent_tmp_dir
> ```
> 
> This will replace the default temporary location with a custom one for all calling instances _while_ this PR is only making sure the vm objects (running vms) are persistent, leaving the default temporary directory for everything else. We could call `get_tmp_dir(persistent=True)` only from the current locations but this brings me to the second point.

Ok, let me explain further, to make sure the vm objects are persistent, there must be a singleton for tracing and managing them and, as you know, it is the `env` object, and then we can let vms be present across test instances.

Generally, `env` can be in-memory objects and could be shared with test instances directly, and in this case you don't really care about how it is implemented because it can work properly. However, since test instances are running within different processes, it is not easy to make env still in-memory, and hence, it is stored in disk currently and the `tmp` dir of test instances is where it is placed (although IMO, the best place should be the `run` dir, but avocado doesn't have this concept so far). 

Now let's take a look at the issue you met, since each test instances (within one avocado/avocado-vt job) will recreate the `tmp` dir, it is not possible to make sure `env` is persistent, and such that the vm objects can't be persistent too. To resolve the issue, the key point is to make sure `env` is persistent, and then the dir where `env` is placed should be persistent (no matter what and how).

So, what I suggested is to let `env` be still in the **`tmp`** dir and make sure the `tmp` dir is persistent, therefore, `data_dir.get_tmp_dir` should point to the same path as well. Otherwise, as what the current approach does, you are attempting to introduce some new directory resource and it just happens to be called "tmp", and it is explained that, about the second point, why you thought the changed code locations don't apply the same logic, that's because we do have two kinds of directory then.

Moreover, not only `env` would be placed in the `tmp` dir, some other temporary resources used by the vms (such as sockets, directories, images) may be placed in the same room, if we have two kinds of `tmp` dir, these temporary resources should be placed in the persistent one too, but other developers may confuse about that.

> 
> 2. The changed code locations don't necessarily apply the same logic. While some (more than one) will check for persistent location first and if not will default to the original temporary directory, some have to additionally check if persistent location is set (in a boolean fashion) and this will require at least two lines of importing `settings` like above at each location which is about the same number of lines of code we try to reduce by refactoring into `get_tmp_dir` on the first place.
> 
> 3. We also have other parameters used strictly for persistent vm objects like "vm_action" which are also Cartesian parameters like the "persistent_tmp_path". All of these parameters are Cartesian since this is the normal way to operate in `env_process.py` and `qemu_vm.py` and after some search I couldn't find a single instance of importing `avocado.settings` or `avocado_vt.options` into any module in the entire `virttest` folder.

Sorry, I just forgot to send other comments together in the first one. About the functionality of `vm_action` my question is, is it really necessary? If we would let some "outdated" vms run in the current frame, does it mean these vms are more likely *requested*? wouldn't these vms be defined in configurations?

Reply:
> Now let's take a look at the issue you met, since each test instances (within one avocado/avocado-vt job) will recreate the `tmp` dir, it is not possible to make sure `env` is persistent, and such that the vm objects can't be persistent too. To resolve the issue, the key point is to make sure `env` is persistent, and then the dir where `env` is placed should be persistent (no matter what and how).

Yes, I confirm everything so far.

> So, what I suggested is to let `env` be still in the **`tmp`** dir and make sure the `tmp` dir is persistent, therefore, `data_dir.get_tmp_dir` should point to the same path as well. Otherwise, as what the current approach does, you are attempting to introduce some new directory resource and it just happens to be called "tmp", and it is explained that, about the second point, why you thought the changed code locations don't apply the same logic, that's because we do have two kinds of directory then.

Do we know the extend to which this will affect other tests and code however? Are we sure that there are no parts of the code that will actually require a temporary storage that should be cleaned up in between tests/runs? I don't want to break other people's code which is why defining an extra optional and more persistent location seemed less intrusive than this. We can make preserving the temporary location optional, but not separately for each part of code, it is either permanent or nonpermanent per run and user configuration.

> Moreover, not only `env` would be placed in the `tmp` dir, some other temporary resources used by the vms (such as sockets, directories, images) may be placed in the same room, if we have two kinds of `tmp` dir, these temporary resources should be placed in the persistent one too, but other developers may confuse about that.

Yes, I see it might be confusing for other developers. If you think making the temporary directory more persistent with a switch should not cause too much trouble we can go that path. Can you confirm this is what you think?

> > 1. The changed code locations don't necessarily apply the same logic. While some (more than one) will check for persistent location first and if not will default to the original temporary directory, some have to additionally check if persistent location is set (in a boolean fashion) and this will require at least two lines of importing `settings` like above at each location which is about the same number of lines of code we try to reduce by refactoring into `get_tmp_dir` on the first place.
> > 2. We also have other parameters used strictly for persistent vm objects like "vm_action" which are also Cartesian parameters like the "persistent_tmp_path". All of these parameters are Cartesian since this is the normal way to operate in `env_process.py` and `qemu_vm.py` and after some search I couldn't find a single instance of importing `avocado.settings` or `avocado_vt.options` into any module in the entire `virttest` folder.
> 
> Sorry, I just forgot to send other comments together in the first one. About the functionality of `vm_action` my question is, is it really necessary? If we would let some "outdated" vms run in the current frame, does it mean these vms are more likely _requested_? wouldn't these vms be defined in configurations?

Yes, the parameter is necessary since our plugin for instance can perform separate setup tests on different vms, say test A for vm1 and test B for vm3 and then can run a conjoint test C using both vm1 and vm3 after the previous test validation. The default action of removing the vm is reasonable for a temporary storage like the one so far but is needed for cases of reuse of previously used vms. I can show you sample code if you are interested.

Reply:
I have pushed the changes you requested - please see below.